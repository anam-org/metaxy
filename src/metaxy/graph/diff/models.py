"""Core data models for graph rendering."""

from enum import Enum
from typing import TYPE_CHECKING, Any

from pydantic import Field
from typing_extensions import Self

from metaxy.models.bases import FrozenBaseModel
from metaxy.models.types import FeatureKey, FieldKey

if TYPE_CHECKING:
    from metaxy.models.feature import FeatureGraph


class NodeStatus(str, Enum):
    """Status of a node in a diff view."""

    NORMAL = "normal"  # Normal node (not in diff mode)
    UNCHANGED = "unchanged"  # Unchanged in diff
    ADDED = "added"  # Added in diff
    REMOVED = "removed"  # Removed in diff
    CHANGED = "changed"  # Changed in diff


class FieldNode(FrozenBaseModel):
    """Represents a field within a feature node.

    Attributes:
        key: Field key
        version: Current field version hash
        old_version: Previous field version hash (for diffs)
        code_version: Code version (if available)
        status: Field status (for diff rendering)
    """

    key: FieldKey
    version: str | None = None  # None if field was removed
    old_version: str | None = None  # For diff mode
    code_version: int | None = None
    status: NodeStatus = NodeStatus.NORMAL


class GraphNode(FrozenBaseModel):
    """Represents a feature node in the graph.

    Attributes:
        key: Feature key
        version: Current feature version hash
        old_version: Previous feature version hash (for diffs)
        code_version: Code version (if available)
        fields: List of field nodes
        dependencies: List of feature keys this node depends on
        status: Node status (for diff rendering)
        project: Project name this feature belongs to
        metadata: Additional custom metadata
    """

    key: FeatureKey
    version: str | None = None  # None if feature was removed
    old_version: str | None = None  # For diff mode
    code_version: int | None = None
    fields: list[FieldNode] = Field(default_factory=list)
    dependencies: list[FeatureKey] = Field(default_factory=list)
    status: NodeStatus = NodeStatus.NORMAL
    project: str | None = None  # Project name (None for legacy features)
    metadata: dict[str, Any] = Field(default_factory=dict)


class EdgeData(FrozenBaseModel):
    """Represents an edge between two nodes.

    Attributes:
        from_key: Source feature key (dependency)
        to_key: Target feature key (dependent)
    """

    from_key: FeatureKey
    to_key: FeatureKey


class GraphData(FrozenBaseModel):
    """Container for complete graph structure.

    This is the unified data model used by all renderers.

    Attributes:
        nodes: Map from feature key string to GraphNode
        edges: List of edges
        snapshot_version: Optional snapshot version
        old_snapshot_version: Optional old snapshot version (for diffs)
    """

    nodes: dict[str, GraphNode]  # Key is feature_key.to_string()
    edges: list[EdgeData] = Field(default_factory=list)
    snapshot_version: str | None = None
    old_snapshot_version: str | None = None  # For diff mode

    def get_node(self, key: FeatureKey) -> GraphNode | None:
        """Get node by feature key.

        Args:
            key: Feature key to lookup

        Returns:
            GraphNode if found, None otherwise
        """
        return self.nodes.get(key.to_string())

    def get_nodes_by_status(self, status: NodeStatus) -> list[GraphNode]:
        """Get all nodes with a specific status.

        Args:
            status: Status to filter by

        Returns:
            List of nodes with matching status
        """
        return [node for node in self.nodes.values() if node.status == status]

    def to_struct(self) -> dict[str, Any]:
        """Serialize to struct (native Python types for storage).

        Note: This uses custom serialization instead of Pydantic's model_dump() because:
        1. Polars struct columns require specific type conversions (e.g., None → "" for strings, None → 0 for ints)
        2. Custom types (FeatureKey, FieldKey) need explicit string conversion for storage
        3. The storage schema is a separate concern from the domain model's Python representation
        4. Different storage backends may need different serialization formats in the future

        Returns:
            Dict with structure compatible with Polars struct type
        """
        nodes_list = []
        for node in self.nodes.values():
            fields_list = []
            for field in node.fields:
                fields_list.append(
                    {
                        "key": field.key.to_string(),
                        "version": field.version if field.version is not None else "",
                        "code_version": field.code_version
                        if field.code_version is not None
                        else 0,
                    }
                )

            nodes_list.append(
                {
                    "key": node.key.to_string(),
                    "version": node.version if node.version is not None else "",
                    "code_version": node.code_version
                    if node.code_version is not None
                    else 0,
                    "fields": fields_list,
                    "dependencies": [dep.to_string() for dep in node.dependencies],
                    "project": node.project if node.project is not None else "",
                }
            )

        edges_list = []
        for edge in self.edges:
            edges_list.append(
                {
                    "from_key": edge.from_key.to_string(),
                    "to_key": edge.to_key.to_string(),
                }
            )

        result: dict[str, Any] = {
            "nodes": nodes_list,
            "edges": edges_list,
        }

        # Include snapshot_version if present
        if self.snapshot_version is not None:
            result["snapshot_version"] = self.snapshot_version

        # Include old_snapshot_version if present (for diffs)
        if self.old_snapshot_version is not None:
            result["old_snapshot_version"] = self.old_snapshot_version

        return result

    @classmethod
    def from_struct(cls, struct_data: dict[str, Any]) -> Self:
        """Deserialize from struct.

        Args:
            struct_data: Dict with structure from to_struct()

        Returns:
            GraphData instance
        """
        nodes = {}
        for node_data in struct_data["nodes"]:
            fields = []
            for field_data in node_data["fields"]:
                fields.append(
                    FieldNode(
                        key=FieldKey(field_data["key"].split("/")),
                        version=field_data["version"]
                        if field_data["version"]
                        else None,
                        code_version=field_data["code_version"]
                        if field_data["code_version"] != 0
                        else None,
                    )
                )

            node = GraphNode(
                key=FeatureKey(node_data["key"].split("/")),
                version=node_data["version"] if node_data["version"] else None,
                code_version=node_data["code_version"]
                if node_data["code_version"] != 0
                else None,
                fields=fields,
                dependencies=[
                    FeatureKey(dep.split("/")) for dep in node_data["dependencies"]
                ],
                project=node_data.get("project") if node_data.get("project") else None,
            )
            nodes[node_data["key"]] = node

        edges = []
        for edge_data in struct_data["edges"]:
            edges.append(
                EdgeData(
                    from_key=FeatureKey(edge_data["from_key"].split("/")),
                    to_key=FeatureKey(edge_data["to_key"].split("/")),
                )
            )

        # Extract snapshot_version if present
        snapshot_version = struct_data.get("snapshot_version")

        # Extract old_snapshot_version if present (for diffs)
        old_snapshot_version = struct_data.get("old_snapshot_version")

        return cls(
            nodes=nodes,
            edges=edges,
            snapshot_version=snapshot_version,
            old_snapshot_version=old_snapshot_version,
        )

    @classmethod
    def from_feature_graph(cls, graph: "FeatureGraph") -> "GraphData":
        """Convert a FeatureGraph to GraphData.

        Args:
            graph: FeatureGraph instance

        Returns:
            GraphData with all nodes and edges
        """
        from metaxy.models.plan import FQFieldKey

        nodes: dict[str, GraphNode] = {}
        edges: list[EdgeData] = []

        # Convert each feature to a GraphNode
        for feature_key, feature_cls in graph.features_by_key.items():
            feature_key_str = feature_key.to_string()
            spec = feature_cls.spec

            # Get feature version
            feature_version = graph.get_feature_version(feature_key)

            # Convert fields
            field_nodes: list[FieldNode] = []
            if spec.fields:
                for field_spec in spec.fields:
                    # Compute field version
                    fq_field_key = FQFieldKey(feature=feature_key, field=field_spec.key)
                    field_version = graph.get_field_version(fq_field_key)

                    field_node = FieldNode(
                        key=field_spec.key,
                        version=field_version,
                        code_version=field_spec.code_version,
                        status=NodeStatus.NORMAL,
                    )
                    field_nodes.append(field_node)

            # Extract dependencies
            dependencies: list[FeatureKey] = []
            if spec.deps:
                dependencies = [dep.key for dep in spec.deps]

            # Get project from feature class
            feature_project = feature_cls.project  # type: ignore[attr-defined]

            # Create node
            node = GraphNode(
                key=feature_key,
                version=feature_version,
                code_version=spec.code_version,
                fields=field_nodes,
                dependencies=dependencies,
                status=NodeStatus.NORMAL,
                project=feature_project,
            )
            nodes[feature_key_str] = node

            # Create edges
            for dep_key in dependencies:
                edges.append(EdgeData(from_key=dep_key, to_key=feature_key))

        return cls(
            nodes=nodes,
            edges=edges,
            snapshot_version=graph.snapshot_version,
        )

    @classmethod
    def from_merged_diff(cls, merged_data: dict[str, Any]) -> "GraphData":
        """Convert merged diff data to GraphData.

        Args:
            merged_data: Merged diff data from GraphDiffer.create_merged_graph_data()

        Returns:
            GraphData with status annotations
        """
        from metaxy.graph.diff.diff_models import FieldChange

        nodes: dict[str, GraphNode] = {}
        edges: list[EdgeData] = []

        # Convert nodes
        for feature_key_str, node_data in merged_data["nodes"].items():
            # Parse feature key
            feature_key = FeatureKey(feature_key_str.split("/"))

            # Map status strings to NodeStatus enum
            status_str = node_data["status"]
            if status_str == "added":
                status = NodeStatus.ADDED
            elif status_str == "removed":
                status = NodeStatus.REMOVED
            elif status_str == "changed":
                status = NodeStatus.CHANGED
            elif status_str == "unchanged":
                status = NodeStatus.UNCHANGED
            else:
                status = NodeStatus.NORMAL

            # Convert fields
            fields_dict = node_data.get("fields", {})
            field_changes_list = node_data.get("field_changes", [])

            # Build field change map for quick lookup
            field_change_map: dict[str, FieldChange] = {}
            for fc in field_changes_list:
                if isinstance(fc, FieldChange):
                    field_change_map[fc.field_key.to_string()] = fc

            # Get all field keys (from both current fields and removed fields in changes)
            all_field_keys = set(fields_dict.keys())
            all_field_keys.update(field_change_map.keys())

            field_nodes: list[FieldNode] = []
            for field_key_str in all_field_keys:
                # Parse field key
                field_key = FieldKey(field_key_str.split("/"))

                # Determine field status and versions
                if field_key_str in field_change_map:
                    fc = field_change_map[field_key_str]
                    if fc.is_added:
                        field_status = NodeStatus.ADDED
                        field_version = fc.new_version
                        old_field_version = None
                    elif fc.is_removed:
                        field_status = NodeStatus.REMOVED
                        field_version = None
                        old_field_version = fc.old_version
                    elif fc.is_changed:
                        field_status = NodeStatus.CHANGED
                        field_version = fc.new_version
                        old_field_version = fc.old_version
                    else:
                        field_status = NodeStatus.UNCHANGED
                        field_version = fc.new_version or fc.old_version
                        old_field_version = None
                else:
                    # Unchanged field
                    field_status = NodeStatus.UNCHANGED
                    field_version = fields_dict.get(field_key_str)
                    old_field_version = None

                field_node = FieldNode(
                    key=field_key,
                    version=field_version,
                    old_version=old_field_version,
                    status=field_status,
                )
                field_nodes.append(field_node)

            # Parse dependencies
            dependencies = [
                FeatureKey(dep_str.split("/"))
                for dep_str in node_data.get("dependencies", [])
            ]

            # Create node
            node = GraphNode(
                key=feature_key,
                version=node_data.get("new_version"),
                old_version=node_data.get("old_version"),
                fields=field_nodes,
                dependencies=dependencies,
                status=status,
            )
            nodes[feature_key_str] = node

        # Convert edges
        for edge_dict in merged_data["edges"]:
            from_key = FeatureKey(edge_dict["from"].split("/"))
            to_key = FeatureKey(edge_dict["to"].split("/"))
            edges.append(EdgeData(from_key=from_key, to_key=to_key))

        return cls(
            nodes=nodes,
            edges=edges,
        )
