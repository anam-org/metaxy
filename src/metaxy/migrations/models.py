"""Data models for migration system."""

from datetime import datetime
from typing import TYPE_CHECKING, Any, Literal

import pydantic

if TYPE_CHECKING:
    from metaxy.migrations.ops import BaseOperation


def _load_operation_class(class_path: str) -> type:
    """Dynamically import operation class.

    Args:
        class_path: Full class path (e.g., "myproject.migrations.S3Backfill")

    Returns:
        Operation class

    Raises:
        ImportError: If module or class not found
    """
    from metaxy.migrations.ops import BaseOperation

    module_path, class_name = class_path.rsplit(".", 1)
    module = __import__(module_path, fromlist=[class_name])
    cls = getattr(module, class_name)

    if not issubclass(cls, BaseOperation):
        raise TypeError(f"{class_path} must be a subclass of BaseOperation, got {cls}")

    return cls


def _register_operation_classes(operations_data: list[dict[str, Any]]) -> None:
    """Register all operation classes for Pydantic discriminated union.

    Dynamically imports user-defined operation classes so Pydantic can
    deserialize them based on the 'type' discriminator field.

    Args:
        operations_data: List of operation dicts from YAML
    """
    for op_data in operations_data:
        op_type = op_data.get("type")
        if not op_type:
            continue

        # Try to import the class (registers it with Pydantic)
        try:
            _load_operation_class(op_type)
        except Exception:
            # Will fail later during validation if class doesn't exist
            # This is just best-effort registration
            pass


class Migration(pydantic.BaseModel):
    """Complete migration definition.

    A migration contains a list of operations that are executed in order.
    Each operation can be any subclass of BaseOperation (e.g.,
    DataVersionReconciliation, MetadataBackfill, or user-defined).

    Operations are polymorphic - deserialized based on the 'type' field.

    Migrations form an explicit dependency chain via parent_migration_id.
    This ensures migrations are applied in the correct order without
    relying on implicit timestamp-based sorting.
    """

    version: int  # Migration schema version
    id: str  # Unique ID (format: "migration_YYYYMMDD_HHMMSS")
    parent_migration_id: str | None = (
        None  # Parent migration (None for first migration)
    )
    from_snapshot_id: (
        str  # Feature graph snapshot before migration (current state in store)
    )
    to_snapshot_id: str  # Feature graph snapshot after migration (target state)
    description: str
    created_at: datetime  # When migration was created

    # Optional overrides for moved/renamed feature classes
    feature_class_overrides: dict[str, str] = pydantic.Field(default_factory=dict)

    # Operations as list of dicts - will be validated/loaded dynamically
    # Can't use Union with discriminator because we support user-defined subclasses
    operations: list[dict[str, Any]]

    def get_operations(self) -> list["BaseOperation"]:
        """Parse operations into BaseOperation objects.

        Returns:
            List of parsed operation objects

        Raises:
            ImportError: If operation type cannot be imported
            ValidationError: If operation data is invalid
        """

        parsed_operations = []

        for op_data in self.operations:
            op_type = op_data.get("type")
            if not op_type:
                raise ValueError(f"Operation missing 'type' field: {op_data}")

            # Load the operation class
            op_class = _load_operation_class(op_type)

            # Parse and validate
            op_instance = op_class.model_validate(op_data)  # type: ignore[attr-defined]
            parsed_operations.append(op_instance)

        return parsed_operations

    @staticmethod
    def from_yaml(path: str) -> "Migration":
        """Load migration from YAML file.

        Supports polymorphic operations - dynamically loads operation classes
        based on the 'type' field.

        Args:
            path: Path to YAML migration file

        Returns:
            Migration object

        Raises:
            FileNotFoundError: If file doesn't exist
            pydantic.ValidationError: If YAML structure is invalid
            ImportError: If operation type class cannot be imported
        """
        import yaml

        with open(path) as f:
            data = yaml.safe_load(f)

        # Register operation classes for polymorphic deserialization
        _register_operation_classes(data.get("operations", []))

        # Pydantic will handle validation and parsing with discriminated unions
        return Migration.model_validate(data)

    def to_yaml(self, path: str) -> None:
        """Save migration to YAML file.

        Args:
            path: Path to write YAML file
        """
        import yaml

        # Use Pydantic's model_dump for serialization
        # Use by_alias=True to export "from" instead of "from_"
        data = self.model_dump(mode="python", by_alias=True)

        with open(path, "w") as f:
            yaml.dump(data, f, default_flow_style=False, sort_keys=False)


class MigrationResult(pydantic.BaseModel):
    """Result of executing a migration."""

    migration_id: str
    status: Literal["completed", "failed", "skipped"]
    operations_applied: int
    operations_failed: int
    affected_features: list[str]
    errors: dict[str, str]  # feature_key -> error message
    duration_seconds: float
    timestamp: datetime

    def summary(self) -> str:
        """Human-readable summary of migration result.

        Returns:
            Multi-line summary string
        """
        lines = [
            f"Migration: {self.migration_id}",
            f"Status: {self.status.upper()}",
            f"Timestamp: {self.timestamp.isoformat()}",
            f"Duration: {self.duration_seconds:.2f}s",
            f"Operations: {self.operations_applied} applied, {self.operations_failed} failed",
            f"Affected features: {len(self.affected_features)}",
        ]

        if self.affected_features:
            lines.append("\nFeatures updated:")
            for feature in self.affected_features:
                lines.append(f"  âœ“ {feature}")

        if self.errors:
            lines.append("\nErrors:")
            for feature, error in self.errors.items():
                lines.append(f"  âœ— {feature}: {error}")

        return "\n".join(lines)
