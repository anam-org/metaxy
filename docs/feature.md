## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
