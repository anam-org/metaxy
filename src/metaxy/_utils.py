import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
