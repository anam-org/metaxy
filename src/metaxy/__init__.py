from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
