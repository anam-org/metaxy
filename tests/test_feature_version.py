"""Tests for feature_version() method."""

from syrupy.assertion import SnapshotAssertion

from metaxy import (
    ContainerDep,
    ContainerKey,
    ContainerSpec,
    Feature,
    FeatureDep,
    FeatureKey,
    FeatureSpec,
)


def test_feature_version_deterministic(snapshot: SnapshotAssertion) -> None:
    """Test that feature_version is deterministic."""

    class TestFeature(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["test", "feature"]),
            deps=None,
            containers=[
                ContainerSpec(key=ContainerKey(["default"]), code_version=1),
            ],
        ),
    ):
        pass

    version1 = TestFeature.feature_version()
    version2 = TestFeature.feature_version()

    # Should be deterministic
    assert version1 == version2

    # Should be 64 characters
    assert len(version1) == 64

    # Should be hex string
    assert all(c in "0123456789abcdef" for c in version1)

    # Snapshot the hash
    assert version1 == snapshot


def test_feature_version_changes_with_code_version(snapshot: SnapshotAssertion) -> None:
    """Test that feature_version changes when code_version changes."""
    from metaxy.models.feature import FeatureRegistry

    # Use separate registries
    registry_v1 = FeatureRegistry()
    registry_v2 = FeatureRegistry()

    class FeatureV1(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["versioned", "feature", "test_v1"]),
            deps=None,
            containers=[
                ContainerSpec(key=ContainerKey(["default"]), code_version=1),
            ],
        ),
        registry=registry_v1,
    ):
        pass

    class FeatureV2(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["versioned", "feature", "test_v2"]),
            deps=None,
            containers=[
                ContainerSpec(
                    key=ContainerKey(["default"]), code_version=2
                ),  # Changed!
            ],
        ),
        registry=registry_v2,
    ):
        pass

    v1 = FeatureV1.feature_version()
    v2 = FeatureV2.feature_version()

    # Should be different (different code_version)
    assert v1 != v2

    # Snapshot both
    assert {"v1": v1, "v2": v2} == snapshot


def test_feature_version_changes_with_dependencies(snapshot: SnapshotAssertion) -> None:
    """Test that feature_version changes when dependencies change."""

    class UpstreamFeature(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["test_deps", "upstream"]),
            deps=None,
            containers=[
                ContainerSpec(key=ContainerKey(["default"]), code_version=1),
            ],
        ),
    ):
        pass

    class DownstreamNoDeps(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["test_deps", "downstream", "no_deps"]),
            deps=None,  # No dependencies
            containers=[
                ContainerSpec(key=ContainerKey(["default"]), code_version=1),
            ],
        ),
    ):
        pass

    class DownstreamWithDeps(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["test_deps", "downstream", "with_deps"]),
            deps=[
                FeatureDep(key=FeatureKey(["test_deps", "upstream"]))
            ],  # Added dependency!
            containers=[
                ContainerSpec(key=ContainerKey(["default"]), code_version=1),
            ],
        ),
    ):
        pass

    no_deps = DownstreamNoDeps.feature_version()
    with_deps = DownstreamWithDeps.feature_version()

    # Should be different
    assert no_deps != with_deps

    # Snapshot both
    assert {"no_deps": no_deps, "with_deps": with_deps} == snapshot


def test_feature_version_multi_container(snapshot: SnapshotAssertion) -> None:
    """Test feature_version with multiple containers."""

    class MultiContainer(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["multi"]),
            deps=None,
            containers=[
                ContainerSpec(key=ContainerKey(["frames"]), code_version=1),
                ContainerSpec(key=ContainerKey(["audio"]), code_version=1),
                ContainerSpec(key=ContainerKey(["metadata"]), code_version=2),
            ],
        ),
    ):
        pass

    version = MultiContainer.feature_version()

    # Should be 64 characters
    assert len(version) == 64

    # Snapshot
    assert version == snapshot


def test_feature_version_with_container_deps(snapshot: SnapshotAssertion) -> None:
    """Test feature_version includes container-level dependencies."""

    class Upstream(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["test_container_deps", "upstream"]),
            deps=None,
            containers=[
                ContainerSpec(key=ContainerKey(["frames"]), code_version=1),
                ContainerSpec(key=ContainerKey(["audio"]), code_version=1),
            ],
        ),
    ):
        pass

    class DownstreamNoDeps(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["test_container_deps", "downstream", "no_deps"]),
            deps=[FeatureDep(key=FeatureKey(["test_container_deps", "upstream"]))],
            containers=[
                ContainerSpec(
                    key=ContainerKey(["default"]),
                    code_version=1,
                    # No container deps
                ),
            ],
        ),
    ):
        pass

    class DownstreamWithContainerDeps(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["test_container_deps", "downstream", "with_deps"]),
            deps=[FeatureDep(key=FeatureKey(["test_container_deps", "upstream"]))],
            containers=[
                ContainerSpec(
                    key=ContainerKey(["default"]),
                    code_version=1,
                    deps=[
                        ContainerDep(
                            feature_key=FeatureKey(["test_container_deps", "upstream"]),
                            containers=[
                                ContainerKey(["frames"]),
                                ContainerKey(["audio"]),
                            ],
                        )
                    ],
                ),
            ],
        ),
    ):
        pass

    no_container_deps = DownstreamNoDeps.feature_version()
    with_container_deps = DownstreamWithContainerDeps.feature_version()

    # Should be different
    assert no_container_deps != with_container_deps

    # Snapshot both
    assert {
        "no_container_deps": no_container_deps,
        "with_container_deps": with_container_deps,
    } == snapshot
