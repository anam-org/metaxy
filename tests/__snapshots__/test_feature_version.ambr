# serializer version: 1
# name: test_feature_version_changes_with_code_version
  dict({
    'v1': '2a68135ecef5cffeb310598c567babad3f7ca34053cf5f51f7aeb34de51d7880',
    'v2': '87daa733b69d31d0cfbbb93a538a33e4742c0eae371618f1284b47cfe285db06',
  })
# ---
# name: test_feature_version_changes_with_dependencies
  dict({
    'no_deps': '8b1eebd5c96addd4e4126b13597f54a0502b646b01579d648ebbf3740f5300e0',
    'with_deps': 'c641fc7efa6288d969142019342fa9b47c266c876a8ea31fd808fdbf8bf18893',
  })
# ---
# name: test_feature_version_deterministic
  '33c004fe4137e09bce3821702bae48ebe3a821183ac3f7e5369efaa177d07857'
# ---
# name: test_feature_version_multi_field
  '76cc968df48602c4cd03d86798001057375f33aae833aea66754838479eb6b07'
# ---
# name: test_feature_version_with_field_deps
  dict({
    'no_field_deps': '9783311bcc336745058386ba6d177000c71a51b0edd716fa3ca8ae7e73e912d6',
    'with_field_deps': '27c14775d3390dbbc397cd841adec9736766098d00f4425238c06e825e8dd8cc',
  })
# ---
